client/src/app/users/user.service.spec.ts

```diff
@@ -59,135 +60,249 @@ describe('User service: ', () => {
-
-    req.flush(testUsers);
-  });
-
-  it('getUsers() calls api/users with filter parameter \'age\'', () => {
-
-    userService.getUsers({ age: 25 }).subscribe(
-      users => expect(users).toBe(testUsers)
-    );
-
-    // Specify that (exactly) one request will be made to the specified URL with the age parameter.
-    const req = httpTestingController.expectOne(
-      (request) => request.url.startsWith(userService.userUrl) && request.params.has('age')
-    );
-
-    // Check that the request made to that URL was a GET request.
-    expect(req.request.method).toEqual('GET');
-
-    // Check that the age parameter was '25'
-    expect(req.request.params.get('age')).toEqual('25');
-
-    req.flush(testUsers);
-  });
-
-  it('getUsers() calls api/users with multiple filter parameters', () => {
-
-    userService.getUsers({ role: 'editor', company: 'IBM', age: 37 }).subscribe(
-      users => expect(users).toBe(testUsers)
-    );
-
-    // Specify that (exactly) one request will be made to the specified URL with the role, company, and age parameters.
-    const req = httpTestingController.expectOne(
-      (request) => request.url.startsWith(userService.userUrl)
-        && request.params.has('role') && request.params.has('company') && request.params.has('age')
-    );
-
-    // Check that the request made to that URL was a GET request.
-    expect(req.request.method).toEqual('GET');
+  describe('When getUsers() is called with parameters, it correctly forms the HTTP request (Javalin/Server filtering)', () => {
+    /*
+    * We really don't care what `getUsers()` returns in the cases
+    * where the filtering is happening on the server. Since all the
+    * filtering is happening on the server, `getUsers()` is really
+    * just a "pass through" that returns whatever it receives, without
+    * any "post processing" or manipulation. So the tests in this
+    * `describe` block all confirm that the HTTP request is properly formed
+    * and sent out in the world, but don't _really_ care about
+    * what `getUsers()` returns as long as it's what the HTTP
+    * request returns.
+    *
+    * So in each of these tests, we'll keep it simple and have
+    * the (mocked) HTTP request return the entire list `testUsers`
+    * even though in "real life" we would expect the server to
+    * return return a filtered subset of the users.
+    */
+
+    it('correctly calls api/users with filter parameter \'admin\'', () => {
+        const mockedMethod = spyOn(httpClient, 'get').and.returnValue(of(testUsers));
+
+        userService.getUsers({ role: 'admin' }).subscribe((users: User[]) => {
+          // The array of `User`s returned by `getUsers()` should be
+          // the array `testUsers`. This is "weird" because we'd truly be expecting
+          // the server to return just `admin` users, but as mentioned above, we're
+          // not trying to get the server here.
+          expect(users)
```

Comment:
Given the comments up above, do we even want to bother checking what `users` gets returned by `getUsers()`? Maybe all we really care about is that it gets called once (line 125) and with the right params (line 132).

The only argument I can see for testing this is to make sure that `getUsers()` does actually return what it gets back from the server, and doesn't modify that list in some way. If that's a reason to keep it, then we'd want to change the comment above this line to say that.

Thoughts?

Reply:
Just to clarify (because it's oddly hard for me to track which things are "the comments above" and be sure I understand what's going on...):

- There are comments in the code about how we are checking that we got back all of the users since the filtering would be taken care of by the server.
- Since we only actually care about talking to the right endpoint and how many times things were called, we could just remove the parts that verify that we got all the users back.
- So... for anything that is only filtered on the server, we could remove the part that is checking something we actually don't care about? (e.g., age and role are only filtered on the server)

Reply:
No, I think it definitely is tricky to make sense of context when you add a comment in a review. All that shows up here is the link you marked and a quite small number of lines above/around it, and that makes it hard to know what things refer to. I'll try to do a better job of copying in more info in comments like this in future.

Your summary is pretty much what I was thinking. We have a bunch of tests where we return, e.g., all the the users, and then check that we return all the users, even though we might reasonably expect some kind of filtering to have happened. Given that in the test we say what the HTTP request should return, it's not clear what the value is of testing that we get it back. As it says somewhere in the comments, about the only reason to check that is to make sure that `getUsers()` or whatever doesn't mess with or manipulate those values.

So I guess the "big" question is whether we want to include those checks or not. I'm really not sure what I think would be the best approach, but it seems that we want to be consistent whatever we choose to do. My (weak) inclination is to leave out those checks since they are mostly just checking "obvious" things, but I can be convinced otherwise.